tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -852,13 +974,193 @@ class SPIRVEmitter : public ASTConsumer {
+
+    // Recursively dencending until we find the true base vector. In the
+    // meanwhile, collecting accessors in the reverse order.
+    *basePtr = expr->getBase();
+    while (const auto *vecElemBase =
+               dyn_cast<HLSLVectorElementExpr>(*basePtr)) {
+      accessors.push_back(vecElemBase->getEncodedElementAccess());
+      *basePtr = vecElemBase->getBase();
+    }
+
+    *flattenedAccessor = accessors.back();
+    for (int32_t i = accessors.size() - 2; i >= 0; --i) {
+      const auto &currentAccessor = accessors[i];
+
+      // Apply the current level of accessor to the flattened accessor of all
+      // previous levels of ones.
+      hlsl::VectorMemberAccessPositions combinedAccessor;
+      for (uint32_t j = 0; j < currentAccessor.Count; ++j) {
+        uint32_t currentPosition = 0;
+        currentAccessor.GetPosition(j, &currentPosition);
+        uint32_t previousPosition = 0;
+        flattenedAccessor->GetPosition(currentPosition, &previousPosition);
+        combinedAccessor.SetPosition(j, previousPosition);
+      }
+      combinedAccessor.Count = currentAccessor.Count;
+      combinedAccessor.IsValid =
+          flattenedAccessor->IsValid && currentAccessor.IsValid;
+
+      *flattenedAccessor = combinedAccessor;
+    }
+  }
+
+  uint32_t doHLSLVectorElementExpr(const HLSLVectorElementExpr *expr) {
+    const Expr *baseExpr = nullptr;
+    hlsl::VectorMemberAccessPositions accessor;
+    condenseVectorElementExpr(expr, &baseExpr, &accessor);
+
+    const QualType baseType = baseExpr->getType();
+    assert(hlsl::IsHLSLVecType(baseType));
+    const auto baseSize = hlsl::GetHLSLVecSize(baseType);
+
+    const uint32_t type = typeTranslator.translateType(expr->getType());
+    const auto accessorSize = accessor.Count;
+
+    // Depending on the number of elements selected, we emit different
+    // instructions.
+    // For vectors of size greater than 1, if we are only selecting one element,
+    // typical access chain or composite extraction should be fine. But if we
+    // are selecting more than one elements, we must resolve to vector specific
+    // operations.
+    // For size-1 vectors, if we are selecting their single elements multiple
+    // times, we need composite construct instructions.
+
+    if (accessorSize == 1) {
+      if (baseSize == 1) {
+        // Selecting one element from a size-1 vector. The underlying vector is
+        // already treated as a scalar.
+        return doExpr(baseExpr);
+      }
+
+      // If the base is an lvalue, we should emit an access chain instruction
+      // so that we can load/store the specified element. For rvalue base,
+      // we should use composite extraction. We should check the immediate base
+      // instead of the original base here since we can have something like
+      // v.xyyz to turn a lvalue v into rvalue.
+      if (expr->getBase()->isGLValue()) { // E.g., v.x;
+        // TODO: select the correct storage class
```

Comment:
We are not handling many kind of variables now. This should be a thing to fix when adding support for global variables and more variables with semantics. #102 for this.